src/transaction/log_recovery_redo.hpp

```diff
@@ -644,6 +645,17 @@ void log_rv_redo_record_sync_apply (THREAD_ENTRY *thread_p, log_rv_redo_context
 			 VPID_AS_ARGS (&rcv_vpid), (int) rcv.offset, (int) rcv.length);
     }
 
+  if (is_passive_transaction_server() && log_data.rcvindex == RVPGBUF_DEALLOC)
+    {
+      /* TODO:
+       * redofunc for RVPGBUF_DEALLOC (pgbuf_rv_dealloc_redo ()) invalidates the page buffer in PTS,
+       * but it does not set page lsa (pgbuf_set_lsa ()) before invalidation.
```

Comment:
The sequence of lines in `pgbuf_rv_dealloc_redo`:

```
pgbuf_rv_dealloc_redo (THREAD_ENTRY * thread_p, const LOG_RCV * rcv)
{
  ...
#if defined(SERVER_MODE)
  if (is_passive_transaction_server ())
    {
      ...
      const int error_code = pgbuf_invalidate (thread_p, rcv->pgptr);
      ...
    }
#endif
   ...
}
```

has been added as part of fixes for http://jira.cubrid.org/browse/LETS-558 and http://jira.cubrid.org/browse/LETS-511 .
See PR #3960 .

Can we revisit that fix with the knowledge of this situation?

> The issue arose from the fact that RVPGBUF_DEALLOC logs, which are not part of the atomic sequence, were causing problems when replicated to the PTS

That fix was aimed to ensure that a page was not mistakenly reused by the PTS while having already gone through the de-allocate - re-allocate sequence.

If there is another way to solve this de-allocate - re-allocate sequence without going through the `pgbuf_invalidate` I think we should revisit the fix done in `pgbuf_rv_dealloc_redo`.

Does a call to `pgbuf_set_lsa` solve the problem?
I think **not** because the set LSA would still be smaller than the required LSA needed upon a subsequent access on PTS.

By looking at `pgbuf_dealloc_page`, it also goes a long way to invalidate the page and make sure that if it is reused, it is correctly reused (ie: re-allocated).
Specifically, I am looking at the comment:
```
  /* how it works: page is "deallocated" by resetting its type to PAGE_UNKNOWN. also prepare bcb for victimization.
   *
   * note: the bcb used to be invalidated. but that means flushing page to disk and waiting for IO write. that may be
   *       too slow. if we add the bcb to the bottom of a lru list, it will be eventually flushed by flush thread and
   *       victimized. */
```

Maybe `pgbuf_rv_dealloc_redo` can do the same on PTS. Instead of just calling `pgbuf_invalidate`.
Can we extract a chunk out of `pgbuf_dealloc_page` (ie: the part after the call to `PGBUF_BCB_LOCK`) and reuse it in `pgbuf_rv_dealloc_redo`?

The reason I ask this is because there are already a lot of "exceptions".
In particular I would want to avoid `if (is_passive_transaction_server()` as much as possible.
With this fix it looks like fixing a workaround (that in #3960 ) with yet another workaround.

Reply:
Thank you, I will add more description in the issue